# 4.9 Intro to Visualisation with Python

## This script contains the following points:

## 1. Import Libraries

## 2. Import Data

## 3. Data Exploration

#### 3.1 Check for Missing Values - Column Check

#### 3.2 Rename Columns

#### 3.3 Explore Data Types of Columns

#### 3.4 Explore Missing Values in df

#### 3.5 Explore Duplicate Values in df

#### 3.6 Data Consistency Check

### 4. Exporting Clean Dataframe

# 1. Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# 2. Import Data

In [2]:
# Creating path string to folder
path = r'C:\Users\Admin\Desktop\Instacart Basket Analysis'

In [3]:
# Importing customers df from csv file
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv')) 

# 3. Data Exploration

In [5]:
# Print head of customers dataframe
customers.head(30)

user_id First Name     Surnam  Gender           STATE  Age date_joined  \
0     26711    Deborah   Esquivel  Female        Missouri   48    1/1/2017   
1     33890   Patricia       Hart  Female      New Mexico   36    1/1/2017   
2     65803    Kenneth     Farley    Male           Idaho   35    1/1/2017   
3    125935   Michelle      Hicks  Female            Iowa   40    1/1/2017   
4    130797        Ann    Gilmore  Female        Maryland   26    1/1/2017   
5    133128    Cynthia      Noble  Female        Kentucky   43    1/1/2017   
6    152052      Chris     Walton    Male         Montana   20    1/1/2017   
7    168851     Joseph    Hickman    Male  South Carolina   30    1/1/2017   
8     69965     Jeremy       Vang    Male           Texas   47    1/1/2017   
9     82820      Shawn      Chung    Male        Virginia   26    1/1/2017   
10    26441     Gloria   Stafford  Female          Nevada   19    1/1/2017   
11    33835      Roger      Casey    Male        Nebraska   45    1/1/2017   
12    75032    Cynthia   Mcdaniel  Female         Georgia   75    1/1/2017   
13       51  Stephanie     Jensen  Female         Wyoming   37    1/1/2017   
14      516      Peter       Hunt    Male        Colorado   51    1/1/2017   
15     6965    Brandon       Mata    Male    North Dakota   70    1/1/2017   
16     8794        Amy       Mack  Female        Nebraska   30    1/1/2017   
17     9944      Brian     Oneill    Male       Wisconsin   26    1/1/2017   
18    13313     Victor       Moss    Male          Alaska   31    1/1/2017   
19    17335    Patrick     Fowler    Male         Vermont   65    1/1/2017   
20    29864      Randy      Mills    Male    North Dakota   39    1/1/2017   
21    37744      Wanda      Salas  Female        Arkansas   77    1/1/2017   
22    50000     Sandra      Trejo  Female           Maine   64    1/1/2017   
23    55567     Martha      Hurst  Female  North Carolina   67    1/1/2017   
24    57549      Shawn      Moore    Male         Montana   72    1/1/2017   
25    60386      Wayne   Woodward    Male          Alaska   59    1/1/2017   
26    68032    Charles  Stevenson    Male   West Virginia   74    1/1/2017   
27    83910       Ruth       Rice  Female         Indiana   70    1/1/2017   
28    86936   Kimberly     Monroe  Female          Oregon   57    1/1/2017   
29    99199   Kimberly     Monroe  Female        Arkansas   80    1/1/2017   

    n_dependants                        fam_status  income  
0              3                           married  165665  
1              0                            single   59285  
2              2                           married   99568  
3              0                            single   42049  
4              1                           married   40374  
5              2                           married   49643  
6              0                            single   61746  
7              0                            single   63712  
8              1                           married  162432  
9              2                           married   32072  
10             2  living with parents and siblings   43443  
11             2                           married   47988  
12             1                           married   57835  
13             1                           married   71190  
14             2                           married  146559  
15             0                  divorced/widowed  100091  
16             3                           married   57096  
17             1                           married   91826  
18             2                           married   91111  
19             1                           married   34514  
20             1                           married   44364  
21             1                           married  125977  
22             2                           married  115489  
23             2                           married   48645  
24             0                  divorced/widowed  135302  

In [6]:
# Print tail of customers dataframe
customers.tail(15)

user_id   First Name      Surnam  Gender           STATE  Age  \
206194   189337        Shawn        Wood    Male      New Jersey   56   
206195   205766       Amanda       Hodge  Female          Oregon   18   
206196   139950       Gloria      Murray  Female        Colorado   45   
206197    74598  Christopher   Velazquez    Male       Minnesota   52   
206198    83573       Gloria      Murray  Female        Michigan   28   
206199   179673         Adam  Villanueva    Male         Wyoming   77   
206200   192637        Kelly    Petersen  Female         Florida   47   
206201    75529        Kathy       Avila  Female   West Virginia   25   
206202   157533          Ned       Stark    Male            Utah   53   
206203   122741         Carl       Weber    Male        Michigan   24   
206204   168073         Lisa        Case  Female  North Carolina   44   
206205    49635       Jeremy     Robbins    Male          Hawaii   62   
206206   135902        Doris    Richmond  Female        Missouri   66   
206207    81095         Rose     Rollins  Female      California   27   
206208    80148      Cynthia       Noble  Female        New York   55   

       date_joined  n_dependants                        fam_status  income  
206194    4/1/2020             1                           married   40373  
206195    4/1/2020             2  living with parents and siblings   48510  
206196    4/1/2020             2                           married  150954  
206197    4/1/2020             0                            single  140700  
206198    4/1/2020             0                            single   32237  
206199    4/1/2020             0                  divorced/widowed  162239  
206200    4/1/2020             0                            single   50638  
206201    4/1/2020             2                           married   56513  
206202    4/1/2020             0                            single   35973  
206203    4/1/2020             2                           married   73357  
206204    4/1/2020             1                           married  148828  
206205    4/1/2020             3                           married  168639  
206206    4/1/2020             2                           married   53374  
206207    4/1/2020             1                           married   99799  
206208    4/1/2020             1                           married   57095

In [7]:
# print shape of dataframe
customers.shape

(206209, 10)

In [8]:
# print info about dataframe
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [13]:
# Print column names of df
customers.columns

Index(['user_id', 'First Name', 'Surnam', 'Gender', 'STATE', 'Age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')

In [14]:
# Print descriptive statistics of df
customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

### 3.1 Check for Missing Values - Column Check

In [9]:
# Frequency check of user_id column
customers['user_id'].value_counts(dropna = False)

2047      1
133652    1
154142    1
147997    1
150044    1
         ..
183539    1
181490    1
187633    1
185584    1
2049      1
Name: user_id, Length: 206209, dtype: int64

In [10]:
# Frequency check of First Name column
customers['First Name'].value_counts(dropna = False)

NaN        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Eugene       197
Merry        197
Garry        191
David        186
Ned          186
Name: First Name, Length: 208, dtype: int64

In [11]:
# Frequency check of Surnam column
customers['Surnam'].value_counts(dropna = False)

Hamilton      252
Randall       248
Lamb          243
Pennington    243
Barnett       242
             ... 
Poole         172
Bauer         166
Pearson       164
Payne         163
Jordan        162
Name: Surnam, Length: 1000, dtype: int64

In [15]:
# Frequency check of Gender column
customers['Gender'].value_counts(dropna = False)

Male      104067
Female    102142
Name: Gender, dtype: int64

In [16]:
# Frequency check of STATE column
customers['STATE'].value_counts(dropna = False)

Arizona                 4044
Illinois                4044
Alabama                 4044
Indiana                 4044
Iowa                    4044
California              4044
Hawaii                  4044
District of Columbia    4044
Delaware                4044
Colorado                4044
Alaska                  4044
Georgia                 4044
Arkansas                4044
Idaho                   4044
Florida                 4044
Connecticut             4044
Kansas                  4043
Mississippi             4043
Oklahoma                4043
Washington              4043
Wyoming                 4043
Tennessee               4043
Louisiana               4043
New York                4043
Utah                    4043
South Dakota            4043
Oregon                  4043
Nevada                  4043
Michigan                4043
New Jersey              4043
New Hampshire           4043
Minnesota               4043
North Dakota            4043
Texas                   4043
Missouri      

In [17]:
# Frequency check of Age column
customers['Age'].value_counts(dropna = False)

19    3329
51    3317
55    3317
56    3306
32    3305
      ... 
65    3145
25    3127
66    3114
50    3102
36    3101
Name: Age, Length: 64, dtype: int64

In [18]:
# Frequency check of date_joined column
customers['date_joined'].value_counts(dropna = False)

9/17/2018     213
2/10/2018     212
9/21/2019     211
4/1/2019      211
12/19/2017    210
             ... 
9/1/2018      141
1/22/2018     140
11/24/2017    139
7/18/2019     138
8/6/2018      128
Name: date_joined, Length: 1187, dtype: int64

In [19]:
# Frequency check of n_dependants column
customers['n_dependants'].value_counts(dropna = False)

0    51602
3    51594
1    51531
2    51482
Name: n_dependants, dtype: int64

In [20]:
# Frequency check of fam_status column
customers['fam_status'].value_counts(dropna = False)

married                             144906
single                               33962
divorced/widowed                     17640
living with parents and siblings      9701
Name: fam_status, dtype: int64

In [21]:
# Frequency check of income column
customers['income'].value_counts(dropna = False)

95891    10
95710    10
57192    10
98675     9
97532     9
         ..
37719     1
45907     1
47954     1
41809     1
90348     1
Name: income, Length: 108012, dtype: int64

I won't delete any columns here as the only column missing a number of values is the 'First Name' column. This column has a total of 11259 NaN values, this represents only 5% of our total data so I have decided to not delete the column.

### 3.2 Rename Columns

In [22]:
# Print column names of df
customers.columns

Index(['user_id', 'First Name', 'Surnam', 'Gender', 'STATE', 'Age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')

In [23]:
# Correct name of Surnam column to Surname
customers.rename(columns = {'Surnam':'Surname'},inplace = True)

In [24]:
# Correct format of STATE column to State
customers.rename(columns = {'STATE':'State'},inplace = True)

In [25]:
# Rename n_dependants column to no_of_dependents
customers.rename(columns = {'n_dependants':'no_of_dependents'},inplace = True)

In [26]:
# Print column names to check changes
customers.columns

Index(['user_id', 'First Name', 'Surname', 'Gender', 'State', 'Age',
       'date_joined', 'no_of_dependents', 'fam_status', 'income'],
      dtype='object')

### 3.3 Explore Data Types of Columns

In [27]:
# Print data types of columns
customers.dtypes

user_id              int64
First Name          object
Surname             object
Gender              object
State               object
Age                  int64
date_joined         object
no_of_dependents     int64
fam_status          object
income               int64
dtype: object

In [28]:
# Change data type of user id column from integer to string
customers['user_id'] = customers['user_id'].astype(str)

In [29]:
# Print data type of user id column
customers['user_id'].dtype

dtype('O')

In [30]:
# Check for mixed data type columns
for col in customers.columns.tolist():
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customers[weird]) > 0:
    print (col)

First Name


In [31]:
# Change data type of First Name column to string
customers['First Name'] = customers['First Name'].astype(str)

In [32]:
# Check for mixed data type columns
for col in customers.columns.tolist():
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customers[weird]) > 0:
    print (col)

In [33]:
# Print data type of First Name column
customers['First Name'].dtype

dtype('O')

### 3.4 Explore Missing Values in df

In [34]:
# Find missing values in df
customers.isnull().sum()

user_id             0
First Name          0
Surname             0
Gender              0
State               0
Age                 0
date_joined         0
no_of_dependents    0
fam_status          0
income              0
dtype: int64

No missing values found

Question - I know from my exploration of the 'First Name' column that I have 11259 NaN values in this column - why are they not picked up when running the above check?

### 3.5 Explore Duplicate Values in df

In [37]:
# Search for duplicated rows in customers and create a subset that contains only these rows
customers_dups = customers[customers.duplicated()]

In [38]:
# Print subset
customers_dups

Empty DataFrame
Columns: [user_id, First Name, Surname, Gender, State, Age, date_joined, no_of_dependents, fam_status, income]
Index: []

In [39]:
# Check shape of subset
customers_dups.shape

(0, 10)

No full duplicate data was found within df customers

### 3.6 Data Consistency Checks

In [40]:
# Print descriptive statistics of df
customers.describe()

Age  no_of_dependents         income
count  206209.000000     206209.000000  206209.000000
mean       49.501646          1.499823   94632.852548
std        18.480962          1.118433   42473.786988
min        18.000000          0.000000   25903.000000
25%        33.000000          0.000000   59874.000000
50%        49.000000          1.000000   93547.000000
75%        66.000000          3.000000  124244.000000
max        81.000000          3.000000  593901.000000

In [41]:
# Explore values within income column that are greater than 125000 using loc() function
customers.loc[customers['income']>125000]


user_id   First Name     Surname  Gender           State  Age  \
0        26711      Deborah    Esquivel  Female        Missouri   48   
8        69965       Jeremy        Vang    Male           Texas   47   
14         516        Peter        Hunt    Male        Colorado   51   
21       37744        Wanda       Salas  Female        Arkansas   77   
24       57549        Shawn       Moore    Male         Montana   72   
...        ...          ...         ...     ...             ...  ...   
206196  139950       Gloria      Murray  Female        Colorado   45   
206197   74598  Christopher   Velazquez    Male       Minnesota   52   
206199  179673         Adam  Villanueva    Male         Wyoming   77   
206204  168073         Lisa        Case  Female  North Carolina   44   
206205   49635       Jeremy     Robbins    Male          Hawaii   62   

       date_joined  no_of_dependents        fam_status  income  
0         1/1/2017                 3           married  165665  
8         1/1/2017                 1           married  162432  
14        1/1/2017                 2           married  146559  
21        1/1/2017                 1           married  125977  
24        1/1/2017                 0  divorced/widowed  135302  
...            ...               ...               ...     ...  
206196    4/1/2020                 2           married  150954  
206197    4/1/2020                 0            single  140700  
206199    4/1/2020                 0  divorced/widowed  162239  
206204    4/1/2020                 1           married  148828  
206205    4/1/2020                 3           married  168639  

[50744 rows x 10 columns]

In [42]:
# Explore values within income column that are greater than 250000 using loc() function
customers.loc[customers['income']>250000]

user_id First Name   Surname  Gender           State  Age date_joined  \
34      117740       Lisa    Sparks  Female          Oregon   55    1/1/2017   
434     159362       Tina   Shannon  Female        Missouri   74    1/3/2017   
818      15683    Michael  Carrillo    Male      New Mexico   40    1/5/2017   
979     200930    Charles   Nichols    Male  South Carolina   60    1/6/2017   
991     136298      Kevin    Ortega    Male      New Mexico   47    1/6/2017   
...        ...        ...       ...     ...             ...  ...         ...   
203425  144788      Henry      Kemp    Male       Wisconsin   64   3/16/2020   
204010   42144      Karen    Parker  Female        Kentucky   75   3/20/2020   
204513   21667     Edward     Miles    Male       Tennessee   53   3/22/2020   
204598    5839    Brandon     Cantu    Male      New Jersey   52   3/23/2020   
206105    5519      Kathy    Daniel  Female         Georgia   78    4/1/2020   

        no_of_dependents fam_status  income  
34                     1    married  292759  
434                    3    married  372334  
818                    1    married  251211  
979                    1    married  300913  
991                    3    married  433206  
...                  ...        ...     ...  
203425                 1    married  400526  
204010                 3    married  297513  
204513                 0     single  518856  
204598                 3    married  275506  
206105                 3    married  262610  

[697 rows x 10 columns]

In [43]:
# Explore values within income column that are greater than 300000 using loc() function
customers.loc[customers['income']>300000]

user_id First Name    Surname  Gender           State  Age date_joined  \
434     159362       Tina    Shannon  Female        Missouri   74    1/3/2017   
979     200930    Charles    Nichols    Male  South Carolina   60    1/6/2017   
991     136298      Kevin     Ortega    Male      New Mexico   47    1/6/2017   
1903     89592  Katherine     Abbott  Female       Minnesota   22   1/12/2017   
2018     33595       Jose  Castaneda    Male     Mississippi   65   1/12/2017   
...        ...        ...        ...     ...             ...  ...         ...   
202655   98580        nan  Patterson    Male      Washington   43   3/12/2020   
202741  100969    Douglas     Campos    Male    Rhode Island   77   3/13/2020   
203226  183344     Louise       Leal  Female       Wisconsin   29   3/15/2020   
203425  144788      Henry       Kemp    Male       Wisconsin   64   3/16/2020   
204513   21667     Edward      Miles    Male       Tennessee   53   3/22/2020   

        no_of_dependents        fam_status  income  
434                    3           married  372334  
979                    1           married  300913  
991                    3           married  433206  
1903                   0            single  317122  
2018                   2           married  577728  
...                  ...               ...     ...  
202655                 2           married  592409  
202741                 0  divorced/widowed  443380  
203226                 3           married  374539  
203425                 1           married  400526  
204513                 0            single  518856  

[382 rows x 10 columns]

In [44]:
# Explore values within income column that are greater than 400000 using loc() function
customers.loc[customers['income']>400000]

user_id First Name    Surname  Gender         State  Age date_joined  \
991     136298      Kevin     Ortega    Male    New Mexico   47    1/6/2017   
2018     33595       Jose  Castaneda    Male   Mississippi   65   1/12/2017   
3074     25696       Sara     Wagner  Female     Tennessee   75   1/18/2017   
5159     79618     Johnny    Cochran    Male   Connecticut   61   1/30/2017   
6515     15931     Willie      Wyatt    Male     Louisiana   48    2/6/2017   
...        ...        ...        ...     ...           ...  ...         ...   
200774   71027   Lawrence     Malone    Male      Michigan   81    3/2/2020   
202655   98580        nan  Patterson    Male    Washington   43   3/12/2020   
202741  100969    Douglas     Campos    Male  Rhode Island   77   3/13/2020   
203425  144788      Henry       Kemp    Male     Wisconsin   64   3/16/2020   
204513   21667     Edward      Miles    Male     Tennessee   53   3/22/2020   

        no_of_dependents        fam_status  income  
991                    3           married  433206  
2018                   2           married  577728  
3074                   0  divorced/widowed  484173  
5159                   1           married  423850  
6515                   0            single  518122  
...                  ...               ...     ...  
200774                 1           married  571932  
202655                 2           married  592409  
202741                 0  divorced/widowed  443380  
203425                 1           married  400526  
204513                 0            single  518856  

[183 rows x 10 columns]

In [45]:
# Explore values within income column that are greater than 500000 using loc() function
customers.loc[customers['income']>500000]

user_id First Name    Surname Gender        State  Age date_joined  \
2018     33595       Jose  Castaneda   Male  Mississippi   65   1/12/2017   
6515     15931     Willie      Wyatt   Male    Louisiana   48    2/6/2017   
12044      658     Ernest    English   Male      Vermont   75   3/11/2017   
21704   136050     Daniel     Walter   Male     Maryland   49    5/6/2017   
21726    58596     Thomas      Cohen   Male   Washington   75    5/6/2017   
...        ...        ...        ...    ...          ...  ...         ...   
199982  169026       Adam  Fernandez   Male       Hawaii   46   2/26/2020   
200450  112113   Benjamin    Parsons   Male      Indiana   57   2/29/2020   
200774   71027   Lawrence     Malone   Male     Michigan   81    3/2/2020   
202655   98580        nan  Patterson   Male   Washington   43   3/12/2020   
204513   21667     Edward      Miles   Male    Tennessee   53   3/22/2020   

        no_of_dependents        fam_status  income  
2018                   2           married  577728  
6515                   0            single  518122  
12044                  0  divorced/widowed  576876  
21704                  3           married  559700  
21726                  2           married  593901  
...                  ...               ...     ...  
199982                 0            single  550917  
200450                 1           married  579397  
200774                 1           married  571932  
202655                 2           married  592409  
204513                 0            single  518856  

[70 rows x 10 columns]

Whilst I thought that the maximum value in the income column seemed high, further exploration of the data leads me to believe this value is reasonable and not the result of an error

# 4. Exporting Clean Dataframe

In [46]:
# Export customers and save to 'Prepared Data'
customers.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'customers_clean.csv'))

Please see new notebook for data merge